src/fnn/fnn.ml

```diff
@@ -586,23 +580,34 @@ module Model = struct
     let inputs ~xs ~ys =
       f_or_d xs_placeholder xs :: f_or_d t.placeholder ys :: addn_inputs
     in
-    for epoch = 1 to epochs do
-      let inputs =
-        match batch_size with
-        | None -> inputs ~xs ~ys
-        | Some batch_size ->
-          let offset = ((epoch-1) * batch_size) % (samples - batch_size) in
+    let batches_per_epoch, batch_inputs =
+      match batch_size with
+      | None -> (1, fun _n -> inputs)
+      | Some batch_size when batch_size >= samples -> (1, fun _step -> inputs)
+      | Some batch_size ->
+        (samples / batch_size,
+        fun step ~xs ~ys ->
+          let offset = (step * batch_size) % (samples - batch_size) in
           let xs = Tensor.sub_left xs offset batch_size in
           let ys = Tensor.sub_left ys offset batch_size in
           inputs ~xs ~ys
-      in
-      let err =
-        Session.run
-          ~inputs
-          ~targets:optimizer
-          (scalar_f_or_d loss)
-      in
-      Stdio.printf "Epoch: %6d/%-6d   Loss: %.2f\n%!" epoch epochs err
+        )
+    in
+    for epoch = 1 to epochs do
+      let err_total = ref 0.0 in
+      for epoch_offset = 0 to batches_per_epoch-1 do
+        let step = (epoch-1) * batches_per_epoch + epoch_offset in
```

Comment:
I would be in favor of just having 'step' equal to 'epoch_offset' here, this would result in the same batches being consistently used between epochs which is likely to be good for debugging reasons.

Reply:
Do you mean that every epoch starts at offset 0 in the training set? This would mean that the last samples in the training set (`samples mod sample_size`) would never be used for training.

Reply:
Yes, indeed the last chunk may not be used but hopefully that's negligible on typical datasets.